### ISSUE TITLE
$sftp->delete(null); doesn't raise an error, but deletes local folders

### ISSUE THREAD

**Issue by magnus-eriksson** [phpseclib/phpseclib#1059]
During development, I had a typo in a variable name, which resulted in me sending in `null` to the `SFTP::delete()` method. Instead of return an error, I noticed that the script (which ran through wp-cli on an Ubuntu box) took an unusually long time.

When I tried again, I got `unknown command wp`. The reason was that the `SFTP::delete()` method had recursively deleted everything in my home folder. My wp-cli bin file was in a sub folder in my home folder.

If course, you should always check what you send in, but in case you do something wrong during development on your local machine, this could be quite bad. I was lucky, since I'm running it in a virtual machine and recently made a snapshot.

_Note:_ In my case, I SFTP'd to the same machine, with the same user I was running the script as (as a mock), but it's still an unexpected behavior.

**Comment by terrafrost** [comment]
It should treat null like ''. So delete would first resolve the path. Assuming the directory hadn't been changed the path it'd resolve to would be /home/user. It'd try to delete that as a file, see that it failed, and then recursively. ie. all the files and all the subdirectories in that folder.

I guess I could make it so that `_realpath` returns a false if the path passed to it wasn't a string. I guess the only potential problem with that would be people passing in objects with `__toString` defined..

**Comment by magnus-eriksson** [comment]
I don't think that an empty string should assume anything on a `delete`-command. I would expect it to fail if a valid filer or folder aren't explicitly given (with an actual value, that is). As I see it, the `delete()`-method should be the equivalent of the `rm`-command, which doesn't work if you haven't explicitly given any file or folder.

To delete files and folders in the current path, it should accept '.' instead.

That's my thoughts and suggestions, anyway. The ramifications of deleting the wrong files and folders might just be too big, otherwise. Specifically since it's a undocumented behavior. 

Btw, I just realized that it shouldn't have deleted any folders at all, since I didn't pass in `true` as a second argument.

**Comment by Sxderp** [comment]
I agree with magnus, the functionality should work similar to rm. If null or empty string is passed then an error should occur.

**Comment by terrafrost** [comment]
That's a good catch about the documentation. I think recursive delete is the more intuitive behavior. You hit delete in any visual SFTP client and it doesn't pop up with a warning saying that you're attempting to delete a directory - it just deletes it and it's content.

Anyway, I've updated the behavior of the delete method:

https://github.com/phpseclib/phpseclib/commit/b70cb8764179060dff71969b1a41f7df76f0b3b6

I'll update the documentation shortly.

> Other than that, I must say that I think this library is awesome. :)

Thanks :)

**Comment by magnus-eriksson** [comment]
A thought, I don't think `empty()` is a good choice in this context, since it also considers `0` and `0.0` to be empty. Having a folder called 0 is not an unlikely scenario. Wouldn't it be better to do something like:

    if (is_bool($path) || trim((string) $path) === '') {
        return false;    
    }

The reason for `is_bool()` is that `true` casts to `1`, which we don't want. The other check will only accept a value that is non empty, for real. so `0`, `0.0` will also be considered a path, which they might be.

Makes sense?

**Comment by bantu** [comment]
is_bool() is a type check

**Comment by bantu** [comment]
The combination of "deleting recursively" and accepting empty values such as null or '' is a rather bad one and should be addressed sooner rather than later. I do not think that using a second boolean parameter on delete() makes sense either. I much cleaner solution in my opinion would be having an extra method for deleting recursively.

**Comment by magnus-eriksson** [comment]
I know. Since the example uses `(string)` to cast the path as string (allows for objects with `__toString()` as they mentioned earlier) and a boolean set to `true` will reutrn `1`, we don't want to accept any booleans as a path.

**Comment by magnus-eriksson** [comment]
>  I much cleaner solution in my opinion would be having an extra method for deleting recursively.

I agree with this.

**Comment by terrafrost** [comment]
```php
if (is_bool($path) || trim((string) $path) === '') {
    return false;    
}
```
I don't agree with the trim() since folders with spaces /can/ be created and since it might, in fact, be desirable to delete them and their contents.

What about this:

```php
if (!is_string($path) || $path == '') {
    return false;
}
```

The issue is that null is converted to '' and so it's appending '' to the current directory. Appending any number of spaces to the current directory doesn't return the current directory - it returns a path in the current directory that has x spaces.

> The combination of "deleting recursively" and accepting empty values such as null or '' is a rather bad one and should be addressed sooner rather than later. I do not think that using a second boolean parameter on delete() makes sense either. I much cleaner solution in my opinion would be having an extra method for deleting recursively.

I don't agree. Quoting my rationale above:

> I think recursive delete is the more intuitive behavior. You hit delete in any visual SFTP client and it doesn't pop up with a warning saying that you're attempting to delete a directory - it just deletes it and it's content.

Same thing in pretty much any GUI. You hit delete in Windows and the directory is deleted. You hit delete in X Windows and the directory is deleted.

**Comment by magnus-eriksson** [comment]
@terrafrost  - Good point about `trim()` and folders with trailing spaces. 

Just checking for `is_string()` will make classes with the magic method `__toString()` fail. 

Maybe something like this:

    if (is_object($path)) {
        // It's an object. Cast it as string before we check anything else. 
        $path = (string) $path;
    }

    if (!is_string($path) || $path == '') {
        return false;
    }
   
This will **only** accept objects and non empty strings. Integers and floats will fail. Maybe it is a reasonable requirement that you cast the path as string yourself before pass it as an argument to this method?

Regarding two different methods or not. I don't think it's really fair to compare this library to a GUI. I have never used a GUI that doesn't make me confirm me deleting a file or a folder.

I rather compare this to command line actions. Sure, you usually don't have different commands for deleting files and folders, but you usually need to explicitly set a flag, (like: `rm` vs `rm -R`) when deleting folders. I think having a switch as a second param is fair enough, but it should be `false` as default. 

Basically, the more changes you make, the more explicit you need to be.

**Comment by terrafrost** [comment]
```php
if (is_object($path)) {
    // It's an object. Cast it as string before we check anything else. 
    $path = (string) $path;
}
```
That'd only work if the object had a `__toString` magic method. What if we do this?:

```php
if (is_object($path) && method_exists($path, '__toString')) {
    // It's an object. Cast it as string before we check anything else. 
    $path = (string) $path;
}
```

> Regarding two different methods or not. I don't think it's really fair to compare this library to a GUI. I have never used a GUI that doesn't make me confirm deleting a file or a folder.

Well in the case of 1.0 and 2.0 changing this behavior, now, would be a possible BC break for apps using phpseclib. For the master branch it's not an issue since the master branch isn't at all concerned with BC, but I'm not really focusing on SFTP / SSH2 changes in that branch atm (I'm more focusing on RSA / DSA / ECDSA / ASN1 stuff). When I do start work on the SSH2 / SFTP changes the BC breaking changes will be numerous.

**Comment by magnus-eriksson** [comment]
@terrafrost - Don't think it's necessary to add the overhead of looking for the `__toString()`-method. Since PHP will throw a fatal error: _"Object of class foo could not be converted to string"_ if it doesn't exist, it might actually help developers debugging their code, in those situations.

I totally get the BC-problem for changing the behavior for recursively deleting folders. 

The documentation should be updated, though, so it says that recursive deletes is the default behavior.

**Comment by terrafrost** [comment]
> Don't think it's necessary to add the overhead of looking for the __toString()-method. Since PHP will throw a fatal error: "Object of class foo could not be converted to string" if it doesn't exist, it might actually help developers debugging their code, in those situations.

That's true. Altho at the same time it seems like if we're gonna go down that path we could also do `user_error('The path that's been passed is not a valid string')` when people pass in null's or int's or whatever. idk. I don't really have any personal preference either way lol.

> The documentation should be updated, though, so it says that recursive deletes are the default behavior.

Yah - that's been done.

**Comment by terrafrost** [comment]
The proposed change has been implemented:

https://github.com/phpseclib/phpseclib/commit/a2d9622bf9644f85644546bb75504a2992756bb6

**Comment by magnus-eriksson** [comment]
Nice. Thank you!
